src/pubsub/mod.rs

```diff
@@ -0,0 +1,395 @@
+                PublishResult::NoSubscribers
+            },
+            |v| {
+                self.metrics.msgs_sent.with_label_values(&["no"]).inc();
+                PublishResult::Success(v)
+            },
+        )
+    }
+}
+
+/// Builder to build a Broker
+pub struct BrokerBuilder<M, T> {
+    opts: Opts,
+    metrics: Metrics,
+    _m: PhantomData<M>,
+    _t: PhantomData<T>,
+}
+
+impl<M: Message, T: TopicId> Default for BrokerBuilder<M, T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<M: Message, T: TopicId> BrokerBuilder<M, T> {
+    /// Create a new BrokerBuilder
+    pub fn new() -> Self {
+        Self {
+            opts: Opts::default(),
+            metrics: Metrics::new(&Registry::new()),
+            _m: PhantomData,
+            _t: PhantomData,
+        }
+    }
+
+    /// Set the max number of topics supported. Default is 1 million.
+    pub const fn with_max_topics(mut self, max_topics: u64) -> Self {
+        self.opts.max_topics = max_topics;
+        self
+    }
+
+    /// Set the idle timeout when an inactive topic is removed. Default it 10min.
+    pub const fn with_idle_timeout(mut self, idle_timeout: Duration) -> Self {
+        self.opts.idle_timeout = idle_timeout;
+        self
+    }
+
+    /// Set per-topic buffer size. Default is 10k.
+    pub const fn with_buffer_size(mut self, buffer_size: usize) -> Self {
+        self.opts.buffer_size = buffer_size;
+        self
+    }
+
+    /// Set per-topic max subscriber limit. Default is 10k.
+    pub const fn with_max_subscribers(mut self, max_subscribers: usize) -> Self {
+        self.opts.max_subscribers = max_subscribers;
+        self
+    }
+
+    /// Set Metrics to use
+    pub fn with_metrics(mut self, metrics: Metrics) -> Self {
+        self.metrics = metrics;
+        self
+    }
+
+    /// Set Prometheus registry to use
+    pub fn with_metric_registry(mut self, registry: &Registry) -> Self {
+        self.metrics = Metrics::new(registry);
+        self
+    }
+
+    /// Build the Broker
+    pub fn build(self) -> Broker<M, T> {
+        Broker::new(self.opts, self.metrics)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_pubsub() {
+        let b: Broker<String, String> = BrokerBuilder::new()
+            .with_buffer_size(3)
+            .with_max_subscribers(1)
+            .build();
+
+        let topic1 = "foo".to_string();
+        let topic2 = "dead".to_string();
+
+        // No subscribers
+        assert_eq!(
+            b.publish(&topic1, "".into()),
+            PublishResult::TopicDoesNotExist
+        );
+        assert_eq!(
+            b.publish(&topic2, "".into()),
+            PublishResult::TopicDoesNotExist
+        );
+        assert_eq!(b.metrics.topics.get(), 0);
+        assert_eq!(b.metrics.msgs_sent.with_label_values(&["yes"]).get(), 2);
+
+        // Subscribe
+        let mut foo_sub = b.subscribe(&topic1).unwrap();
```

Comment:
With this semantics a topic is created with the first subscription and only subscribers can create new topics, not publishers. Maybe this is a correct approach, but it doesn't feel explicit. If everyone unsubscribes the topic is deleted?

Reply:
ah, just saw it at the end of the test